Code/Framework/AzCore/AzCore/Memory/AllocatorInstance.h

```diff
@@ -32,23 +32,49 @@ namespace AZ::AllocatorStorage
                 {
                     m_allocator = Environment::CreateVariable<Allocator>(AzTypeInfo<Allocator>::Name());
                 }
+#if defined(CARBONATED)
+                s_AllocatorEnvironmentVariable = &m_allocator;
+#endif
             }
 
+#if defined(CARBONATED)
+            ~AllocatorEnvironmentVariable()
+            {
+                s_AllocatorEnvironmentVariable = nullptr;
+            }
+#endif
+
             Allocator& operator*() const
             {
                 return *m_allocator;
             }
 
+#if defined(CARBONATED)
+            inline static EnvironmentVariable<Allocator>* s_AllocatorEnvironmentVariable{};
+#endif
         private:
             EnvironmentVariable<Allocator> m_allocator;
         };
 
     public:
+#if defined(CARBONATED)
+        static bool HasAllocator()
+        {
+            static AllocatorEnvironmentVariable s_allocator;
+            return AllocatorEnvironmentVariable::s_AllocatorEnvironmentVariable != nullptr;
+        }
```

Comment:
Could you please explain/detail what's going on here?

Reply:
O3DE has only allocator reference getter, but if calling it late in app life cycle the allocator can be destroyed.
So Alexey had to introduce a boolean checker function to call it prior to get the reference.
Because he calls the checker fn before the getter fn it makes sense to insert the static variable in the checker fn.

Reply:
OK, I understand what's happening now.  s_AllocatorEnvironmentVariable is used as a marker of sorts to check to see if s_allocator has been created (or destroyed).

I don't like having to call "HasAllocator" as it has side effects, in that the name _implies_ a simple check, but the first time it's called, it'll create the allocator itself.  And calling "HasAllocater" in GetAllocator _looks_ like a NOOP unless you dig deeper.

There's already an existing method to check if m_allocator has been constructed, so we can clean this up a bit. See `operator EnvironmentVariable::bool()`

Let's do this:

1. Remove "s_AllocatorEnvironmentVariable" entirely, and use `operator EnvironmentVariable::bool` instead.
2. Add a method to `AllocatorEnvironmentVariable` to check `(bool)m_allocator` instead.
3. Change `HasAllocator` to call (2).  

Something like this:

```
// AllocatorEnvironmentVariable
~AllocatorEnvironmentVariable()
{
    m_allocator.Reset();
}

bool IsConstructed() const
{
       return (bool)m_allocator;
}

// AllocatorStorage
static bool HasAllocator()
{
    return AllocatorEnvironmentVariable::IsConstructed();
}

static IAllocator& GetAllocator()
{
    // Note: Keep the static in GetAllocator
    static AllocatorEnvironmentVariable s_allocator;
    return *s_allocator;
}
```

Reply:
In the game code we have this sequence:
if (!HasAllocator()) return;
GetAllocator().allocate();

What constructs the allocator? In Alexey's code it is in HasAllocator(), which then return true. You insist on HasAllocator() without side effects like allocator creation, so it must be constructed elsewhere before HasAllocator() call like:
TryToCreateAllocator();
if (!HasAllocator()) return;
GetAllocator().allocate();

That does not look nice to me. Maybe we just rename HasAllocator to TryToCreateAllocator, which returns bool on success?
if (!TryToCreateAllocator()) return;
GetAllocator().allocate();

Reply:
In the original implement, GetAllocator does the allocation.  Alexey moved it to HasAllocator.  I'm saying keep the original implementation as is, but implement the checks to see if the allocator is, in fact, still constructed.

Reply:
This works for app's end of life use time, I'm about the app's beginning of life use time. With the first ever use of an allocator it must be constructed before HasAllocator() call, because otherwise HasAllocator() returns false.

Reply:
> This works for app's end of life use time, I'm about the app's beginning of life use time. With the first ever use of an allocator it must be constructed before HasAllocator() call, because otherwise HasAllocator() returns false.

Right, as it should be. "HasAllocator" should return false if it's not constructed.

The "rule" here cannot be that "HasAllocator" must return true always. What it must do is return whether or not the allocator has been created, so we can safely get rid of it later, in allocator.cpp

Which brings us back to "HasAllocator" should return true/false and have no other side effects. It should also use the existing functionality of EnvironmentVariable to see whether or not the object has been constructed, instead of introducing a marker pointer that is set to null at the dtor.